File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function dfs(root: TreeNode | null): [number, number] {
	if (!root) {
		return [0, 0]
	}
	const [left_d, left_l] = dfs(root.left)
	const [right_d, right_l] = dfs(root.right)
	const depth = 1 + Math.max(left_d, right_d)
	const length = Math.max(left_l, right_l, left_d + right_d)
	return [depth, length]
}

function diameterOfBinaryTree(root: TreeNode | null): number {
	return dfs(root)[1]
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root):
        if not root:
            return 0, 0  # depth, length

        left_d, left_l = self.dfs(root.left) if root.left else (0, 0)
        right_d, right_l = self.dfs(root.right) if root.right else (0, 0)
        
        dep = 1 + max(left_d, right_d)
        length = max(left_l, right_l, left_d + right_d)
        return dep, length
        
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        return self.dfs(root)[1]